team-bimmer/ribbit
# Internal css

### WHAT?
- Injecting critical CSS internally into each component's static html file as a single style tag. Critical CSS being only the CSS that the page uses, thus server side rendering a fully styled shell of each react component for the user to serve as a static file. 

### WHY? 
- The benefit of making a CSR single-paged app isomorphic/universal is so that visitors to a website can instantly view the shell of the webpage as feed back that they are at the correct website while the functionality loads (esp on a poor connection). Therefore, it is also essential that this shell be fully styled. Anything less than a fully styled shell of the react component would be poor UX. 

### HOW?
1. During ribbit build, our CLI tool bundles the user's webapp. Within ribbit's webpack.config file, the following lines have been added to extract all CSS into one single file entitled styles.min.css. 
- At the top of the ribbit config file outside of module.exports: 
`const ExtractTextPlugin = require("extract-text-webpack-plugin");
const extractCSS = new ExtractTextPlugin("styles.min.css");`
- Within the webpack config, add the following rule:
`      {
        test: /\.css$/,
        use: extractCSS.extract(["css-loader", "postcss-loader"])
      }`
- Within module.exports, add the following plugin: 
`plugins: [extractCSS]`
2. Within the htmlTemplate middleware, the bundled file with all CSS is passed into the library function purifyCSS along with the reactDOM. For each route, the critical CSS will be saved to a variable that will then be injected into a single style tag within the html template. Specific steps: 
- At the top of the htmlTemplate file, import the purifyCSS library: 
`const purifyCSS = require('purify-css');`
- Within the function htmlTemplate, insert the following code within the body of reactStream.on's callback: 
`const css = fs.readFile(path.resolve(__dirname, '../../dist/styles.min.css'), 'utf8', (err, data) => {
        const criticalCSS = purifyCSS(reactDom, data);
        const html = `
          <!DOCTYPE html>
          <html>
          <head>
          <meta charset="utf-8">
          <title>React SSR</title>
          <style>${criticalCSS}</style>
          </head>
          <body>
          <div id="app">${reactDom}</div>
          <script src="${ribbitConfig.bundleRoot}.js"></script>
          </body>
          </html>
          `;
        res.locals = {
          ...res.locals,
          html,
          appParentDirectory,
          componentRoute
        };
      next();
    })`
### HOW TO TEST? 
- Within a react app using CSS modules, run ribbit build. Then open static files in ribbit.statics to see if page has critical CSS injected as a single style tag. Open static files in browser locally to see that the correct corresponding styles have been applied. 

### FEEDBACK?
